packages/kbn-i18n/src/core/i18n.js

```diff
@@ -0,0 +1,240 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/**
+ @typedef Messages - messages tree, where leafs are translated strings
+ @type {object<string, object>}
+ @property {string} [locale] - locale of the messages
+ @property {object} [formats] - set of options to the underlying formatter
+ */
+
+import IntlMessageFormat from 'intl-messageformat';
+import IntlRelativeFormat from 'intl-relativeformat';
+import memoizeIntlConstructor from 'intl-format-cache';
+import { isString, isObject, hasValues, deepMerge } from './helper';
+import { formats as EN_FORMATS } from './formats';
+
+// Add all locale data to `IntlMessageFormat`.
+// TODO: Use dynamic import for asynchronous loading of specific locale data
+import './locales';
+
+const EN_LOCALE = 'en';
+const LOCALE_DELIMITER = '-';
+const messages = {};
+const getMessageFormat = memoizeIntlConstructor(IntlMessageFormat);
+
+let defaultLocale = EN_LOCALE;
+let currentLocale = EN_LOCALE;
```

Comment:
We should avoid storing state at the module level like this. Instead of a bunch of seemingly static exports that actually mutate global state, could this module export a class that encapsulates the state?

Reply:
Initially I18n was a class, but after the [comment](https://github.com/elastic/kibana/pull/19555#discussion_r193818562) I decided to remove the class and made i18n a singleton object. I don't see real use cases for multiple I18n instances and also multiple instance of I18n can conflict between themselves because all of them will share `IntlMessageFormat` instance and `IntlMessageFormat.defaultLocale` value in particular.

Reply:
@epixa can you please give us a little bit of context on the main reason why you want to avoid global state in this particular case?

I mean, I definitely agree with you in general, but in this case even if we don't use global state in our own code we'll still depend on global mutable state of `IntlMessageFormat` and `IntlRelativeFormat` (e.g. `.defaultLocale` Maksim mentioned, that's unfortunate :/) that may give consumers wrong impression about the state involved and lead to even more confusion/issues....

We could potentially add some guards around that 3rd-party "state mutators" and throw, but it will also look weird.

/cc @archanid

Reply:
From what I can see, `IntlMessageFormat` and `IntlRelativeFormat` have to be used this way, mutating their static properties. We never expose these constructors to the outside world; they're meant to be used within kbn-i18n, which only has a single i18n object. I support this approach.

Reply:
I think this is fine for now, so no need to change it. We can iterate on it later if this becomes a problem.

It is possible to build non-global APIs that themselves rely on global singletons. Every API is different, so there's no silver bullet here, but it's doable. For example (I haven't actually looked into whether this would be a good idea or not), one could set/rollback the global APIs upon every usage within the abstractions we build.

In any case, my biggest concern isn't that the global APIs are handled in this way, but rather that we say "well, there is some globalness here, so let's just make _all_ of our own state related to this global as well". If we followed that logic up and down the stack, then anything that touches document/window/storage would also be a global singleton, which essentially puts web development back about 15 years.